### PR TITLE
feat(step_3): add users and authentication

### PR DESCRIPTION
**Pull request by AlexMomot-717**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -43,7 +43,7 @@ repos:
     rev: 7.0.0
     hooks:
       - id: flake8
-        exclude: task_manager/settings.py
+        exclude: "(task_manager/settings|task_manager/users/forms).py"
```
Comment (svntmr) [inline .pre-commit-config.yaml]: is it necessary to ignore these files?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -57,11 +57,9 @@ repos:
         language: system
         # entry should be the same you'd use to run mypy locally
         entry: mypy task_manager --config-file pyproject.toml
-        pass_filenames: false # not needed, as mypy should be executed on the whole codebase
+#        exclude: task_manager/settings.py
```
Comment (svntmr) [inline .pre-commit-config.yaml]: can you please revert this change?

File: task_manager/users/admin.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import django_stubs_ext
+from django.contrib import admin
+from task_manager.users.models import ServiceUser
+
+django_stubs_ext.monkeypatch()
```
Comment (svntmr) [inline task_manager/users/admin.py]: it's better to do this monkeypatching only once in file like `settings.py`, like [here](https://github.com/wemake-services/wemake-django-template/blob/5bf1569e2710e11befc6991893f94419136d74bd/%7B%7Bcookiecutter.project_name%7D%7D/server/settings/__init__.py#L14-L19)

File: task_manager/users/tests.py
Comment (svntmr) [inline task_manager/users/tests.py:1]: can we remove this file for now?

File: task_manager/users/views.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+#     user_id = kwargs.get("id")
+#     user = ServiceUser.objects.get(id=user_id)
+#     if request.user != user:
+#         message = "У вас нет прав для изменения другого пользователя."
+#         messages.add_message(request, messages.ERROR, message)
+#         return redirect("users_list")
+#     return {
+#         "user_id": user_id,
+#         "user": user
+#     }
```
Comment (svntmr) [inline task_manager/users/views.py]: could you please remove this block?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (AlexMomot-717) [reply .pre-commit-config.yaml:46]: Снимок экрана от 2024-06-29 19-09-23

**Comment by svntmr** [comment]
does it make sense to delete the onrender link from the README if we won't use it anymore?

could you also please add instructions on how to start the project locally?

File: locale/ru/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -32,22 +32,152 @@ msgstr "Менеджер задач"
...
+msgid "You are signed in"
+msgstr "Вы залогинены"
+
+#: task_manager/views.py:30
+msgid ""
+"Please enter the correct username and password. Both fields may be case "
+"sensitive."
+msgstr ""
+"Пожалуйста, введите правильные имя пользователя и пароль. Оба поля могут "
+"быть чувствительны к регистру."
```
Comment (svntmr) [inline locale/ru/LC_MESSAGES/django.po]: wdyt about this option if they're actually case-sensitive?
```suggestion
msgid ""
"Please enter the correct username and password. Both fields are case "
"sensitive."
msgstr ""
"Пожалуйста, введите правильные имя пользователя и пароль. Оба поля "
"чувствительны к регистру."
```

File: task_manager/settings.py [hunk 6]
```diff
@@ -88,34 +95,45 @@
...
+    }
+
 
 # Password validation
 # https://docs.djangoproject.com/en/5.0/ref/settings/#auth-password-validators
 
 AUTH_PASSWORD_VALIDATORS = [
     {
-        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
+        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",  # noqa: E501
```
Comment (svntmr) [inline task_manager/settings.py]: is it necessary to add noqa here? is there an other way to fix it?

Comment (svntmr) [reply task_manager/settings.py]: I think we're missing `.flake8` file with black-compatible configuration

File: task_manager/users/urls.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    UserFormEditView,
+    UserListView,
+)
+
+urlpatterns = [
+    path("", UserListView.as_view(), name="users_list"),
+    path("create/", UserFormCreateView.as_view(), name="create_user"),
+    path("<int:id>/edit/", UserFormEditView.as_view(), name="update_user"),
+    path("<int:id>/delete/", UserFormDeleteView.as_view(), name="delete_user"),
+]
```
Comment (svntmr) [inline task_manager/users/urls.py:14]: do we miss `login` and `logout` urls?

Comment (AlexMomot-717) [reply task_manager/users/urls.py:14]: `login` and `logout` urls specified in the task_manager/urls.py

> Роуты текущего шага
GET /users/ — страница со списком всех пользователей
GET /users/create/ — страница регистрации нового пользователя
POST /users/create/ — создание пользователя
GET /users/<int:pk>/update/ — страница редактирования пользователя
POST /users/<int:pk>/update/ — обновление пользователя
GET /users/<int:pk>/delete/ — страница удаления пользователя
POST /users/<int:pk>/delete/ — удаление пользователя
GET /login/ — страница входа
POST /login/ — аутентификация (вход)
POST /logout/ — завершение сессии (выход)

File: .flake8
Comment (svntmr) [inline .flake8:1]: why not just copy the `.flake8` from the previous project?

File: README.md [hunk 8]
```diff
@@ -3,6 +3,24 @@
...
+### Setup:
 
-https://task-manager-nj02.onrender.com
+Install dependencies:
+
+```
+poetry install
+```
+
+Launch database:
```
Comment (svntmr) [inline README.md]: wdyt about this option?
```suggestion
Build and execute migrations:
```

File: README.md [hunk 9]
```diff
@@ -3,6 +3,24 @@
...
+
+```
+make migrations
+```
+
```
Comment (svntmr) [inline README.md:23]: do I need to create `.env` file?

File: locale/ru/LC_MESSAGES/django.po [hunk 10]
```diff
@@ -32,22 +32,152 @@ msgstr "Менеджер задач"
...
+msgid "You are not authorized! Please sign in."
+msgstr "Вы не авторизованы! Пожалуйста, выполните вход."
+
+#: task_manager/users/views.py:47
+msgid "User signed up successfully."
+msgstr "Пользователь успешно зарегистрирован"
+
+#: task_manager/users/views.py:61 task_manager/users/views.py:92
+msgid "You do not have permission to update or delete another user."
+msgstr "У вас нет прав для изменения другого пользователя."
```
Comment (svntmr) [inline locale/ru/LC_MESSAGES/django.po]:
```suggestion
msgid "You do not have permission to update or delete another user."
msgstr "У вас нет прав для изменения или удаления другого пользователя."
```

File: task_manager/users/templates/create_user.html [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  {% load django_bootstrap5 %}
+  {% load i18n %}
+
+  <div class="container wrapper flex-grow-1">
+    <h1 class="my-4">{% trans "Signing up" %}</h1>
+
+    <form method="post">
+      {% csrf_token %}
+      {% bootstrap_form form %}
+      <input
```
Comment (svntmr) [inline task_manager/users/templates/create_user.html]: this form has fewer requirements in the hexlet example, see:

hexlet - 
image

your app - 
image

**Review by svntmr — COMMENTED** [review]
I'm sorry I completely forgot about this part 🙈:
> Напишите тесты для CRUD пользователей, где C — это регистрация, U — обновление и D — удаление

File: README.md (hunk 9 above)
Comment (AlexMomot-717) [reply README.md:23]: you do not need

Comment (svntmr) [reply README.md:23]: I need, actually 

```py
django.core.exceptions.ImproperlyConfigured: settings.DATABASES is improperly configured. Please supply the ENGINE value. Check settings documentation for more details.
make: *** [migrations] Error 1
```

this happens as `DEBUG` is `False` by default

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -29,6 +29,8 @@ repos:
         # syntax to notify a line break.
         # See https://github.com/pre-commit/pre-commit-hooks/issues/57
         exclude_types: [markdown]
+        pass_filenames: false
```
Comment (svntmr) [inline .pre-commit-config.yaml]: this hook is only executed on the files that were provided to it, so it makes sense to remove `pass_filenames: false`

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -60,13 +62,12 @@ repos:
...
         types: [python, pyi]
-  - repo: local
-    hooks:
       - id: htmlformatter
         name: prettier html formatter
         language: node
         additional_dependencies:
           - prettier@3.3.2
           - prettier-plugin-jinja-template@1.4.0
+        exclude: task_manager/tests/fixtures/
```
Comment (svntmr) [inline .pre-commit-config.yaml]: do we need it if we have `.prettierignore` now?

File: pyproject.toml [hunk 14]
```diff
@@ -16,15 +16,24 @@ whitenoise = {extras = ["brotli"], version = "^6.6.0"}
 gunicorn = "^22.0.0"
 uvicorn = "^0.29.0"
 django-bootstrap5 = "^24.2"
+freezegun = "^1.5.1"
+django-hint = "^0.3.2"
```
Comment (svntmr) [inline pyproject.toml]: are they the main project dependencies? if not, could you please move them to the `dev` group?

File: task_manager/tests/test_user.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+    # when
+    response = client.get(route)
+
+    # then
+    # csrfmiddlewaretoken is different each time
+    assert (
+        "Для подтверждения введите, пожалуйста, пароль ещё раз."
+        in response.content.decode()
+    )
+    assert response.status_code == 200
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: I guess we're missing a test case when user is logged in

File: task_manager/tests/test_user.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+    # given
+    user_id = service_user.id
+    route = f"/users/{user_id}/edit/"
+
+    # when
+    client.force_login(service_user)
+    response = client.get(route)
+
+    # then
+    #  csrfmiddlewaretoken is different each time
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: could you please check if it's possible to set it once for the tests?

File: task_manager/users/views.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+
+class UserFormEditView(UserLoginRequiredMixin, View):
+    def get(self, request: HttpRequest, **kwargs: int | str) -> HttpResponse:
+        user_id = kwargs.get("id")
+        user = get_object_or_404(ServiceUser, pk=user_id)
+        if request.user != user:
+            message = _(
+                "You do not have permission to update or delete another user."
+            )  # noqa: E501
```
Comment (svntmr) [inline task_manager/users/views.py]: [opional] can we do something to remove this noqa?

File: task_manager/users/views.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+from django.views import View
+from task_manager.settings import LOGIN_URL
+from task_manager.users.forms import ServiceUserForm
+from task_manager.users.models import ServiceUser
+
+
+class UserLoginRequiredMixin(LoginRequiredMixin):
+    def dispatch(
+        self, request: HttpRequest, *args: Any, **kwargs: str | int
+    ) -> Any:  # noqa
```
Comment (svntmr) [inline task_manager/users/views.py]: [opional] can we do something to remove this noqa?

File: task_manager/users/views.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+
+class UserFormDeleteView(UserLoginRequiredMixin, View):
+    def get(self, request: HttpRequest, **kwargs: int | str) -> HttpResponse:
+        user_id = kwargs.get("id")
+        user = get_object_or_404(ServiceUser, pk=user_id)
+        if request.user != user:
+            message = _(
+                "You do not have permission to update or delete another user."
+            )  # noqa: E501
```
Comment (svntmr) [inline task_manager/users/views.py]: [opional] can we do something to remove this noqa?

File: task_manager/tests/test_user.py (hunk 15 above)
Comment (AlexMomot-717) [reply task_manager/tests/test_user.py]: it's probably "authenticated_user / unauthenticated_user"  that makes some a mess.
All casses are covered by tests.
It'll be better to change to logged, I suppose

File: .github/workflows/tests.yml [hunk 20]
```diff
@@ -0,0 +1,26 @@
+name: tests
+
+on:
+  - pull_request
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
Comment (svntmr) [inline .github/workflows/tests.yml]: [optional] wdyt, does updating to `v4` make sense?

File: task_manager/tests/test_user.py [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+from django.test import Client
+from freezegun import freeze_time
+from task_manager.tests.paths import resolve_path
+from task_manager.users.forms import ServiceUserForm
+from task_manager.users.models import ServiceUser
+
+pytestmark = pytest.mark.django_db
+
+
+def test_users_list(client: Client, service_user: ServiceUser) -> None:
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: wdyt, does it make sense to include test that has some users pre-created?

File: task_manager/tests/test_user.py [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+        "username": "username",
+        "password1": "password",
+        "password2": "password",
+    }
+
+    # when
+    response = client.post(route, data=form_data)
+
+    # then
+    assert ServiceUser.objects.count() == 1
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: wdyt, does it make sense to check that user was created with the right username?

File: task_manager/tests/test_user.py [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does it make sense to check where users will be redirected to?

File: task_manager/tests/test_user.py [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+    csrf_token = response.context.get("csrf_token")
+    form = ServiceUserForm(instance=service_user)
+    context = {"form": form, "user_id": user_id, "csrf_token": csrf_token}
+
+    # then
+    expected_html = TemplateResponse(
+        response.wsgi_request, "update_user.html", context=context
+    ).render()
+    assert expected_html.content == response.content
+    assert "Изменение пользователя" in response.content.decode()
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: do we need this part if we compare response contents?

File: task_manager/tests/test_user.py [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+) -> None:
+    # given
+    user_id = service_user.id
+    route = f"/users/{user_id}/edit/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does it make sense to check where users will be redirected to?

File: task_manager/tests/test_user.py [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+    # given
+    client.force_login(service_user)
+    object_user_id = ServiceUser.objects.create_user("username", "", "password").id
+    route = f"/users/{object_user_id}/edit/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does it make sense to check where users will be redirected to?

File: task_manager/tests/test_user.py [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+
+
+def test_edit_user_objectuser_not_exists(
```
Comment (svntmr) [inline task_manager/tests/test_user.py]:
```suggestion
def test_edit_user_objectuser_does_not_exist(
```

File: task_manager/tests/test_user.py [hunk 28]
```diff
@@ -0,0 +1,238 @@
...
+        "password1": "password",
+        "password2": "password",
+    }
+
+    # when
+    response = client.post(route, data=form_data)
+
+    # then
+    assert ServiceUser.objects.get(id=1).username == "username"
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does checking where users will be redirected make sense?

File: task_manager/tests/test_user.py [hunk 29]
```diff
@@ -0,0 +1,238 @@
...
+        "password1": "",
+        "password2": "",
+    }
+
+    # when
+    response = client.post(route, data=form_data)
+
+    # then
+    assert ServiceUser.objects.get(id=1).username == user_username
+    assert response.status_code == 200
```
Comment (svntmr) [inline task_manager/tests/test_user.py:197]: why does it return `200_OK` when validation has failed?

File: task_manager/tests/test_user.py [hunk 30]
```diff
@@ -0,0 +1,238 @@
...
+    context = {"user": service_user, "csrf_token": csrf_token}
+
+    # then
+    expected_html = TemplateResponse(
+        response.wsgi_request, "delete_user.html", context=context
+    ).render()
+
+    # then
+    assert expected_html.content == response.content
+    assert (
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: do we need this part if we compare response contents to expected HTML?

File: task_manager/tests/test_user.py [hunk 31]
```diff
@@ -0,0 +1,238 @@
...
+) -> None:
+    # given
+    user_id = service_user.id
+    route = f"/users/{user_id}/delete/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does checking where users will be redirected make sense?

File: task_manager/tests/test_user.py [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+
+
+def test_delete_user_objectuser_not_exists(
```
Comment (svntmr) [inline task_manager/tests/test_user.py]:
```suggestion
def test_delete_user_objectuser_does_not_exist(
```

File: task_manager/tests/test_user.py [hunk 33]
```diff
@@ -0,0 +1,238 @@
...
+    user_id = service_user.id
+    route = f"/users/{user_id}/delete/"
+
+    # when
+    client.force_login(service_user)
+    response = client.post(route)
+
+    # then
+    assert ServiceUser.objects.count() == 0
+    assert response.status_code == 302
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: [optional] wdyt, does checking where users will be redirected make sense?

File: task_manager/users/forms.py [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+
+    class Meta:
+        model = ServiceUser
+        fields = (
+            "first_name",
+            "last_name",
+            "username",
+            "password1",
+            "password2",
+        )  # noqa: E501
```
Comment (svntmr) [inline task_manager/users/forms.py]: do we need this `noqa`?

**Comment by svntmr** [comment]
footer with the `Hexlet` link should be aligned with the left side

image

**Comment by svntmr** [comment]
first name and last name fields are required on the hexlet version of the app

image

**Comment by svntmr** [comment]
login page keeps username in the hexlet implementation

image

**Comment by svntmr** [comment]
login alert and other alerts take less space on the page

image

**Comment by svntmr** [comment]
delete confirmation message uses first and last name instead of username
image

File: task_manager/tests/test_user.py (hunk 21 above)
Comment (AlexMomot-717) [reply task_manager/tests/test_user.py]: you're right this test doesn't present or check any case of user CRUD

File: task_manager/tests/test_user.py (hunk 29 above)
Comment (AlexMomot-717) [reply task_manager/tests/test_user.py:197]: cause the same form is refreshed/renewed if form data is not valid

File: task_manager/tests/test_user.py [hunk 35]
```diff
@@ -0,0 +1,272 @@
...
+    # when
+    response = client.get(route)
+
+    # then
+    context = {"users": users}
+    expected_html = TemplateResponse(
+        response.wsgi_request, "users_list.html", context=context
+    ).render()
+    assert response.content.decode() == expected_html.content.decode()
+    assert "ilon mask" in response.content.decode()
```
Comment (svntmr) [inline task_manager/tests/test_user.py]: do we need this part if we compare response contents?

**Review by svntmr — APPROVED** [review]
